Change flattened dependency graph from `DiGraph<NodeId>` to `DiGraph<SystemKey>`

ItsDoot opened the pull request:
# Objective

- Part of #20115

Semantically, the fully flattened dependency graph never contains `SystemSetKey`s, so lets encode that into its type.

## Solution

- Added `GraphNodeId` trait.
- Generalized `DiGraph` and `UnGraph` with a new `GraphNodeId` `N` type parameter.
- Generalized most functions involving `DiGraph`/`UnGraph` to take a `GraphNodeId` type parameter.
- Added `Graph::try_into` function to help us convert from `DiGraph<NodeId>` to `DiGraph<SystemKey>`. Does it look a bit gnarly? Yea.

## Testing

Re-using current tests.

ItsDoot commented on crates/bevy_ecs/src/schedule/graph/graph_map.rs:
```diff
@@ -249,28 +248,65 @@ where
+    /// # Errors
+    ///
+    /// If the conversion fails, it returns an error of type `T::Error`.
+    pub fn try_into<T: GraphNodeId + TryFrom<N>>(self) -> Result<Graph<DIRECTED, T, S>, T::Error>
```
Suggestions welcome on making this function a bit more legible (it works and I'm lazy).

chescock replied:
Would it help to reframe it as `map`?  Part of the complexity here is bubbling up the conversion errors just to unwrap them with `unreachable!`.  If you instead passed a closure, then you could put the panic in there and not deal with error handling here.  

Otherwise, my best suggestion is to do a little more type inference as `.collect::<Result<_, T::Error>>()?`.  Rust needs help on the error type because `?` does `into()`, but it can infer the success type.

ItsDoot replied:
I've cleaned it up a bit with some inner functions, docs, and used the `_` as suggested. Should I take it further?

chescock commented on crates/bevy_ecs/src/schedule/graph/mod.rs:
```diff
@@ -82,24 +82,24 @@ pub(crate) fn row_col(index: usize, num_cols: usize) -> (usize, usize) {
 
 /// Stores the results of the graph analysis.
-pub(crate) struct CheckGraphResults {
+pub(crate) struct CheckGraphResults<Id: GraphNodeId> {
```
Very minor nit, but you used `N: GraphNodeId` above and `Id: GraphNodeId` here and it might be good to use a consistent type parameter name.

ItsDoot replied:
Aligned them all to `N`.

chescock commented on crates/bevy_ecs/src/schedule/graph/node.rs:
```diff
@@ -1,42 +1,62 @@
+    fn new(id: Self::Id, direction: Direction) -> Self;
+
+    /// Unpacks a [`GraphNodeId`] and a [`Direction`] from this type.
+    fn unwrap(self) -> (Self::Id, Direction);
```
`unwrap()` is usually used for panicking methods.  Would it make sense to call these `pack` and `unpack`?

alice-i-cecile replied:
Agreed; this needs to be changed.

ItsDoot replied:
Done.

chescock commented on crates/bevy_ecs/src/schedule/graph/graph_map.rs:
```diff
@@ -249,28 +248,65 @@ where
+                        .collect::<Result<Vec<T::Directed>, T::Error>>()?,
+                ))
+            })
+            .collect::<Result<IndexMap<T, Vec<T::Directed>, S>, T::Error>>()?;
```
This will do something weird if multiple keys `try_into()` the same value.  We don't need to do that, so it's not worth supporting, but it might be worth noting in the doc comment that it's a requirement.

ItsDoot replied:
Note added in the function docs.

chescock commented on crates/bevy_ecs/src/schedule/graph/node.rs:
```diff
@@ -1,42 +1,62 @@
+}
+
+/// Types that are a pair of [`GraphNodeId`]s.
+pub trait GraphNodeIdPair: Copy + Eq + Hash + Debug {
```
This is probably overengineering, but these two traits are similar and my instinct is to combine them as `trait Packed<T1, T2> { fn new(T1, T2) -> Self; fn unwrap(self) -> (T1, T2) }`.  

Or maybe even get rid of them both and do `type Directed: From<(Self, Direction)> + Into<(Self, Direction)>`.

alice-i-cecile commented on crates/bevy_ecs/src/schedule/graph/graph_map.rs:
```diff
@@ -11,24 +11,23 @@ use core::{
 /// For example, an edge between *1* and *2* is equivalent to an edge between
 /// *2* and *1*.
-pub type UnGraph<S = FixedHasher> = Graph<false, S>;
+pub type UnGraph<N, S = FixedHasher> = Graph<false, N, S>;
```
The N generic needs docs throughout.

ItsDoot commented:
Thoughts on `GraphNodeId` now? I removed the two extra related traits as suggested by @chescock and also renamed the associated types to match their semantics.

@alice-i-cecile @chescock

alice-i-cecile reviewed (approved):
I appreciate the explictness about systems vs system sets. I think this is a net improvement as a result.

chescock commented on crates/bevy_ecs/src/schedule/schedule.rs:
```diff
@@ -1140,29 +1157,19 @@ impl ScheduleGraph {
-            if ambiguous_with_flattened.contains_edge(a, b)
-                || self.ambiguous_with_all.contains(&a)
-                || self.ambiguous_with_all.contains(&b)
+            if ambiguous_with_flattened.contains_edge(a.into(), b.into())
```
This `into()` is the same as the `NodeId::System` on the next line, right?  No preference for which way we do it, but it feels like they should do it the same way.  

```suggestion
            if ambiguous_with_flattened.contains_edge(NodeId::System(a), NodeId::System(b))
```

chescock commented on crates/bevy_ecs/src/schedule/node.rs:
```diff
@@ -251,6 +255,186 @@ new_key_type! {
+
+impl From<CompactNodeIdPair> for (NodeId, NodeId) {
+    fn from(value: CompactNodeIdPair) -> Self {
+        let a = match value.is_system_a {
```
Oh, I didn't notice the `match bool` before.  It was like that before, and it does seem reasonable here for symmetry with the reverse impl, but ... but ... `if`!
